### ISSUE TITLE
malloc error in ClTracerLensing when setting A_SPLINE_MAX < 1

### ISSUE THREAD

**Issue by tilmantroester** [LSSTDESC/CCL#354]
When creating a ClTracerLensing object with A_SPLINE_MAX set to less than 1, Python gives a malloc error. 
```
python(26540,0x7fffb026f3c0) malloc: *** mach_vm_map(size=18446744056529686528) failed (error code=3)
*** error: can't allocate region
*** set a breakpoint in malloc_error_break to debug
ERROR: Could not allocate memory for linear-spaced array (N=-2147483647)
```
The exception is just a type error
```
TypeErrorTraceback (most recent call last)
<ipython-input-2-eabf94f2e53d> in <module>()
     18 
     19 lens1 = ccl.ClTracerLensing(ccl_cosmology, has_intrinsic_alignment=False,
---> 20                             n=[z_nofz, nofz],)

/Users/yooken/miniconda/envs/python3/lib/python3.5/site-packages/pyccl/cls.py in __init__(self, cosmo, has_intrinsic_alignment, n, z, bias_ia, f_red)
    193                  has_intrinsic_alignment=has_intrinsic_alignment,
    194                  z=z, n=n, bias=None, mag_bias=None,
--> 195                  bias_ia=bias_ia, f_red=f_red)
    196 
    197 

/Users/yooken/miniconda/envs/python3/lib/python3.5/site-packages/pyccl/cls.py in __init__(self, cosmo, tracer_type, has_rsd, has_magnification, has_intrinsic_alignment, z, n, bias, mag_bias, bias_ia, f_red, z_source)
     98                             self.z_rf, self.rf,
     99                             self.z_source,
--> 100                             status )
    101 
    102     def __del__(self):

TypeError: 'int' object is not iterable
```

Here is  the code to cause this error, with A_SPLINE_MAX=0.98:
```
import os
import numpy as np
import pyccl as ccl

os.environ["CCL_PARAM_FILE"] = "./ccl_params_modified.ini"

ccl_cosmology = ccl.Cosmology(Omega_c=0.3,
                              Omega_b=0.05,
                              Omega_k=0.0,
                              h=0.7,
                              n_s=0.96,
                              sigma8=0.8)

z_nofz = np.linspace(0.1, 4, 1000)
nofz = np.exp(-0.5*(z_nofz-1.0)**2/0.2**2)

lens1 = ccl.ClTracerLensing(ccl_cosmology, has_intrinsic_alignment=False,
                            n=[z_nofz, nofz],)
```

Tested on master with Python 3.5.3 and 3.6.1.

Upon digging a bit through the code, it seems that ccl_comoving_radial_distance returns NAN for a = 1./(1+zmax), which results in a non-sensical nchi (line 309, ccl_cls.c). Directly calling ccl.comoving_radial_distance with the same a is fine, however.

**Comment by tilmantroester** [comment]
It turns out that setting A_SPLINE_MAX to less than 1 breaks a lot of things because in quite a few places functions get called with a=1. For example
 - ccl_h_over_h0 (isn't this always 1 for a=1?)
https://github.com/LSSTDESC/CCL/blob/f6f2d48e0a7653538a155cf02bf7982f53236eb6/src/ccl_cls.c#L164
 - ccl_linear_spacing https://github.com/LSSTDESC/CCL/blob/f6f2d48e0a7653538a155cf02bf7982f53236eb6/src/ccl_cls.c#L310
 - ccl_linear_matter_power https://github.com/LSSTDESC/CCL/blob/f6f2d48e0a7653538a155cf02bf7982f53236eb6/src/ccl_power.c#L1572

I originally intended to use A_SPLINE_MAX to change the range of integration in the Limber integral but considering how much would have to be changed, I think this can be easier achieved by hacking ccl_angular_cl directly.

Since so many places explicitly use a=1 (often for good reason), does it make sense to remove A_SPLINE_MAX as a user-changeable parameter?

**Comment by c-d-leonard** [comment]
@tilmantroester Did this ever get resolved?

**Comment by tilmantroester** [comment]
No, unfortunately not. The options are to go through the whole code and find instances where a=1 is used in some way and replace that with `A_SPLINE_MAX` (which might break other things) or remove `A_SPLINE_MAX` as a user-settable option.

**Comment by c-d-leonard** [comment]
Okay, thanks for the update. I was just going through older issues to check; looks like we should keep this one open then.

**Comment by c-d-leonard** [comment]
I think A_SPLINE_MAX is no longer a user-facing parameter - if so is it fair to close this? @tilmantroester

**Comment by tilmantroester** [comment]
It still is: https://ccl.readthedocs.io/en/latest/source/notation_and_other_cosmological_conventions.html#controlling-splines-and-numerical-accuracy
There seems to be only few places where `A_SPLINE_MAX` is still used. I'll try to see if replacing all of them with `a=1` works, in which case we can remove the option altogether.

**Comment by c-d-leonard** [comment]
Thank you! Let me know how it goes.

**Comment by tilmantroester** [comment]
Addressed with #751.
